docs/source/notes/extending.func.rst

```diff
@@ -0,0 +1,479 @@
+  (if it exists) staticmethods must be transformable via :func:`torch.vmap`. That
+  is, they must consist of only PyTorch operations (as opposed to e.g. NumPy or custom
+  CUDA kernels).
+
+Example::
+
+    class MyCube(torch.autograd.Function):
+        # Set generate_vmap_rule to True to ask PyTorch to automatically generate
+        # a vmap rule.
+        generate_vmap_rule = True
+
+        @staticmethod
+        def forward(x):
+            result = x ** 3
+            dx = 3 * x ** 2
+            return result, dx
+
+        @staticmethod
+        def setup_context(ctx, inputs, output):
+            x, = inputs
+            result, dx = output
+            ctx.save_for_backward(x, dx)
+
+        @staticmethod
+        def backward(ctx, grad_output, grad_dx):
+            x, dx = ctx.saved_tensors
+            result = grad_output * dx + grad_dx * 6 * x
+            return result
+
+    def my_cube(x):
+        result, dx = MyCube.apply(x)
+        return result
+
+    x = torch.randn(3)
+    result = torch.vmap(my_cube)(x)
+    assert torch.allclose(result, x ** 3)
+
+
+Defining the vmap staticmethod
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If your :meth:`~torch.autograd.function.FunctionCtx.save_for_backward` calls into another system (like NumPy, C++, CUDA, triton),
+then to get it to work with :func:`torch.vmap` or transforms that use it, you'll
+need to manually define a ``vmap`` staticmethod.
+
+Depending on what transforms you want to use and your use case, you may not need
+to add a ``vmap`` staticmethod to all of your :class:`torch.autograd.Function`:
+
+- For example, :func:`torch.func.jacrev` performs :func:`~torch.vmap` over the backward pass.
+  So if you're only interested in using :func:`torch.func.jacrev`, only
+  the :meth:`~Function.backward` staticmethod needs to be vmappable.
+
+We do recommend adding a ``vmap`` staticmethod to all of your :meth:`~torch.autograd.function.FunctionCtx.save_for_backward`
+though, especially if you are writing a third-party library and you want your
+:meth:`~torch.autograd.function.FunctionCtx.save_for_backward` to work with all combinations of :func:`torch.func` transforms.
+
+Conceptually, the vmap staticmethod is responsible for defining how the :meth:`~Function.forward`
+should behave under :func:`torch.vmap`. That is, it defines how to transform
+the :meth:`~Function.forward` to run over inputs with an additional dimension (the dimension
+being vmapped over). This is similar to how :func:`torch.vmap` is implemented over
+PyTorch operations: for each operation, we define a vmap rule (sometimes also
+referred to as a "batching rule").
+
+Here's how to define the ``vmap`` staticmethod:
+
+- the signature is ``vmap(info, in_dims: Tuple[Optional[int]], *args)``, where
+  ``*args`` is the same as the args to :meth:`~Function.forward`.
```

Comment:
should vmap take `**kwargs` for consistency since forward does (though they have to be passed positionally because apply doesn't take true kwargs)

Reply:
I might take this as a follow-up item. I am not sure:
- `setup_context(ctx, inputs, output)` takes in `inputs` that were directly passed to forward
- If there were ANY unspecified values that are default kwargs, then we have a problem -- they're not passed to setup_context, so setup_context doesn't get their default value (if it needs to save them for backward)
- So the conclusion is that autograd.Function with setup_context has some really awkward handling for default kwargs.

I'm inclined to say in the docs that autograd.Function with setup_context doesn't support `**kwargs`

Reply:
I end up changing the documentation to suggest that default args and keywords args be passed via wrapper so it's clearer how to handle them, but not changing the suggested signature of the vmap staticmethod